### PR TITLE
[FIX] account: account: fiscal pos differents bill/delivery addresses

### PR DESCRIPTION
**Pull request by guva-odoo**
We did not compute fiscal position correctly regarding
billing/delivery addresses combinations.
Before this commit, we used the delivery to compute the
fiscal position, except for invoices with no delivery or in the same
country as the company, but it is not sufficient for several use cases.
With this commit we check multiple combinations of
EU/extra-EU partners with/without VAT.

Change has been made regarding this spreadsheet:
https://docs.google.com/spreadsheets/d/1PDlXpsgRiGvYvrUKguJqWPVl-g-7H6YqJiWoRNgWCYI/edit?gid=0#gid=0

opw-4072691

### CODE REVIEW

**Comment by tsb-odoo** [comment]
Hi @epictete, tested and LGTM functionally speaking. 
Will also solve by the way the case of Task 4169910 (the proper way, not the way asked on the ticket)

**Review by epictete — COMMENTED** [review]
Not an easy one to review :sweat_smile: 
Would be nice to have the specific use cases that we are addressing here (i.e. steps to reproduce the problem).
I attempted to clean the code a bit in a fixup commit, let me know what you think :wink:

**Comment by guva-odoo** [comment]
> Would be nice to have the specific use cases that we are addressing here (i.e. steps to reproduce the problem).

Please look at the test, it lists all the use cases ;-)

**Comment by epictete** [comment]
> Please look at the test, it lists all the use cases ;-)

Thanks for the advice, I didn't think of that... :roll_eyes: 
Let me rephrase then.
Having a commit message that says: "It did not work, but with this new commit it works!" is not very helpful.
Which is why we have [guidelines](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#commit-message-full-description) on how to write a good commit message :wink: 
It is especially important when the code lacks clarity and/or comments :nerd_face:

**Review by epictete — COMMENTED** [review]
Thanks for the update, LGTM, just a few details :ok_hand: 
Also, please update your commit message explaining **why** it did not work before and/or **why** it works now (see previous comment) :wink:

File: addons/account/models/partner.py [hunk 1]
```diff
@@ -241,25 +241,39 @@ def _prepare_fpos_base_domain(self, vat_required):
         ]
 
     @api.model
-    def _get_fiscal_position(self, partner, delivery=None):
+    def _get_fiscal_position(self, partner, delivery=None, i=None):
```
[the remark this answers is not in the document]
Comment (guva-odoo) [reply addons/account/models/partner.py:244]: oopsie

File: addons/account/models/partner.py [hunk 2]
```diff
@@ -249,17 +249,34 @@ def _get_fiscal_position(self, partner, delivery=None):
...
+        eu_vat_partner = partner.vat and partner.vat[:2] in eu_country_codes
+        eu_partner = partner.country_code in eu_country_codes
+        eu_delivery = delivery.country_code in eu_country_codes
+        domestic_delivery = delivery_country == company.country_id
+        external_delivery = delivery_country != partner.country_id
+
+        # If the delivery is to a different country than the partner's country (external delivery),
+        # the delivery is within the EU, and the partner does not have a valid EU VAT number,
+        # then assign the company's country as the delivery country) and force vat_valid to True
+        # to get the domestic fp
```
Comment (epictete) [inline addons/account/models/partner.py:270]:
```suggestion
        # then assign the company's country as the delivery country and force vat_valid to True
        # in order to get the domestic FP
```

File: addons/account/models/partner.py [hunk 3]
```diff
@@ -249,17 +249,35 @@ def _get_fiscal_position(self, partner, delivery=None):
...
+        if external_delivery and eu_delivery and not eu_vat_partner:
+            delivery_country = company.country_id
+            vat_valid = True
+
+        # If the delivery is to the same country than the company's country (domestic delivery),
```
Comment (epictete) [inline addons/account/models/partner.py:275]:
```suggestion
        # If the delivery is to the same country as the company's country (domestic delivery),
```

File: addons/account/models/partner.py [hunk 4]
```diff
@@ -249,17 +249,35 @@ def _get_fiscal_position(self, partner, delivery=None):
...
+        # the partner has a valid EU VAT number but is not from EU
+        # We need to force vat_valid to False in order to get the EU private fp
```
Comment (epictete) [inline addons/account/models/partner.py:277]:
```suggestion
        # the partner has a valid EU VAT number that is not from EU,
        # we need to force vat_valid to False in order to get the EU private FP
```

**Review by epictete — COMMENTED** [review]
Maybe we should target 16.0?

File: addons/account/models/partner.py (hunk 4 above)
Comment (guva-odoo) [reply addons/account/models/partner.py:261]: but is the word I want to use, partner has a valid EU VAT, but he is not from EU ;-)

**Review by epictete — APPROVED** [review]
OK, let's go :ok_hand: 

@robodoo r+

**Review by epictete — APPROVED** [review]
@robodoo r+ rebase-ff
